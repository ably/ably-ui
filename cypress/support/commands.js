Cypress.Commands.add(
  "compareComponents",
  ({
    reactComponentUrl,
    viewComponentUrl,
    beforeScreenshot = () => {},
    selector,
    componentName = Cypress.mocha.getRunner().suite.ctx.test.parent.title,
  }) => {
    let reactScreenshotPath = null;
    let vwScreenshotPath = null;

    cy.exec(`mkdir -p cypress/screenshots/component-parity/results`, {
      log: false,
    });

    const diffPath = `cypress/screenshots/component-parity/results/${componentName}.png`;

    // If you are wondering why this isn't Promise.all or async/await have a look at
    // https://docs.cypress.io/guides/core-concepts/introduction-to-cypress.html#Commands-Are-Not-Promises

    cy.visit(reactComponentUrl);

    beforeScreenshot();

    cy.get(selector)
      .screenshot(`${componentName}.react`, {
        onAfterScreenshot: (_, props) => (reactScreenshotPath = props.path),
      })
      .then(() => {
        cy.visit(viewComponentUrl);
        beforeScreenshot();

        cy.get(selector)
          .screenshot(`${componentName}.vw`, {
            onAfterScreenshot: (_, props) => (vwScreenshotPath = props.path),
          })
          .then(() => {
            cy.exec(
              `cypress/support/diff.js ` +
                `--pathA="${reactScreenshotPath}" ` +
                `--pathB="${vwScreenshotPath}" ` +
                `--target="${diffPath}" ` +
                `--threshold="0.005" `
            ).then((result) => {
              cy.exec(`rm "${reactScreenshotPath}" "${vwScreenshotPath}"`, {
                log: false,
              });

              assert.isTrue(result.stdout === "Yay", "Screenshots match");
            });
          });
      });
  }
);
