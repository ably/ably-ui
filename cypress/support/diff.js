#! /usr/bin/env node

// Used by the compareComponents command

const Blink = require("blink-diff");
const argv = require("yargs").argv;

const diff = new Blink({
  imageAPath: argv.pathA,
  imageBPath: argv.pathB,
  imageOutputPath: argv.target,
  threshold: 0.05,
});

diff.run((error, result) => {
  // We read stdout/stderr - this script is executed interactively by cypress
  if (error) {
    console.error(error);
  } else {
    console.log(diff.hasPassed(result.code) ? "Yay" : "Nay");
  }
});
