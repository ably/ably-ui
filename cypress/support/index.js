// ***********************************************************
// This example support/index.js is processed and
// loaded automatically before your test files.
//
// This is a great place to put global configuration and
// behavior that modifies Cypress.
//
// You can change the location of this file or turn off
// automatically serving support files with the
// 'supportFile' configuration option.
//
// You can read more here:
// https://on.cypress.io/configuration
// ***********************************************************

import "cypress-plugin-snapshots/commands";
import "./commands";

// Cypress has a bug that breaks snapshot tests when running the whole suite
// Follow https://github.com/cypress-io/cypress/issues/3090
// Recommended workaround below:
export const fixSnapshotSpec = (filename) => () => {
  const path = require("path");
  // const relative = filename.substr(1); // removes leading "/"
  const projectRoot = Cypress.config("projectRoot");
  const absolute = path.join(projectRoot, filename);

  Cypress.spec = {
    absolute,
    name: path.basename(filename),
    filename,
  };
};
