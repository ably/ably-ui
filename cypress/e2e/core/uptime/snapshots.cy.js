import { fixSnapshotSpec } from "../../../support";

beforeEach(fixSnapshotSpec(__filename));

describe("Uptime Snapshot Test", () => {
  const url = "/components/uptime";
  const sharedSpecs = (url) => {
    beforeEach(() => cy.visit(url));

    it("test case common to both", () => {
      cy.get('[data-id="uptime"]').first().toMatchSnapshot();
      cy.get('[data-id="uptime-error"]').first().toMatchSnapshot();
    });
  };

  beforeEach(() => {
    cy.viewport("macbook-16");
  });

  describe("react", () => {
    sharedSpecs(url);
  });
});
