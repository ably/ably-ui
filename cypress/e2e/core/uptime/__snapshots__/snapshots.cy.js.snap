exports[`Uptime Snapshot Test > react > test case common to both #0`] = `
<div class="ui-uptime-widget" data-id="uptime">
  <div class="ui-uptime-graph">
    <ul class="ui-uptime-list">
      <li class="ui-uptime-day ui-uptime-trouble ui-uptime-align-left">
        <span
          class="ui-uptime-width-extra-wide"
          style="transform: translateX(0%);"
          >09pm for 13 mins - Network connectivity issues in the Asia-Pacific
          region</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span class="ui-uptime-width-normal" style="transform: translateX(0%);"
          >12 May UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-10%);"
          >11 May UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-20%);"
          >10 May UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-30%);"
          >09 May UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-40%);"
          >08 May UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-50%);"
          >07 May UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-trouble ui-uptime-align-left">
        <span
          class="ui-uptime-width-extra-wide"
          style="transform: translateX(-50%);"
          >09pm for 13 mins - Network connectivity issues in the Asia-Pacific
          region</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">05 May UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">04 May UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">03 May UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">02 May UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">01 May UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">30 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">29 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">28 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">27 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">26 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">25 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">24 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">23 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">22 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(50%);"
          >21 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(50%);"
          >20 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(40%);"
          >19 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(30%);"
          >18 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(20%);"
          >17 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(10%);"
          >16 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(0%);"
          >15 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-seperator"></li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span class="ui-uptime-width-normal" style="transform: translateX(0%);"
          >14 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-10%);"
          >13 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-20%);"
          >12 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-30%);"
          >11 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-40%);"
          >10 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-50%);"
          >09 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-50%);"
          >08 Apr UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">07 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">06 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">05 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">04 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">03 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">02 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">01 Apr UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">31 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">30 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">29 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">28 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">27 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">26 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">25 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">24 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">23 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(50%);"
          >22 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(50%);"
          >21 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(40%);"
          >20 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(30%);"
          >19 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(20%);"
          >18 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(10%);"
          >17 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-trouble ui-uptime-align-right">
        <span
          class="ui-uptime-width-extra-wide"
          style="transform: translateX(0%);"
          >Extreme - Ad ipisicing officia be atae rerum soluta cupid it ate
          Libero nulla n…</span
        >
      </li>
      <li class="ui-uptime-seperator"></li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span class="ui-uptime-width-normal" style="transform: translateX(0%);"
          >15 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-10%);"
          >14 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-20%);"
          >13 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-30%);"
          >12 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-40%);"
          >11 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-50%);"
          >10 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-left">
        <span
          class="ui-uptime-width-normal"
          style="transform: translateX(-50%);"
          >09 Mar UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">08 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">07 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">06 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">05 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">04 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">03 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">02 Mar UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-down ui-uptime-align-null">
        <span class="ui-uptime-width-extra-wide"
          >09pm for 13 mins - Network connectivity issues in the Asia-Pacific
          region</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">28 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">27 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">26 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">25 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">24 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">23 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">22 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-null">
        <span class="ui-uptime-width-normal">21 Feb UTC</span>
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(50%);"
          >20 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(50%);"
          >19 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(40%);"
          >18 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(30%);"
          >17 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(20%);"
          >16 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(10%);"
          >15 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(0%);"
          >14 Feb UTC</span
        >
      </li>
      <li class="ui-uptime-day ui-uptime-up ui-uptime-align-right">
        <span class="ui-uptime-width-normal" style="transform: translateX(0%);"
          >13 Feb UTC</span
        >
      </li>
    </ul>
  </div>
  <div class="ui-uptime-footer">
    <div class="ui-uptime-legend">
      <span class="ui-uptime-key ui-uptime-up text-white">All good</span
      ><span class="ui-uptime-key ui-uptime-trouble text-white"
        >Performance issue</span
      ><span class="ui-uptime-key ui-uptime-down text-white"
        >Service disruption</span
      >
    </div>
    <div class="ui-uptime-link-back">
      <a
        href="//status.ably.com"
        class="font-sans font-bold block text-gui-default hover:text-gui-hover focus:text-gui-focus focus:outline-gui-focus group text-p2 py-8 "
        style="--featured-link-icon-size: var(--fs-p2);"
        >More on our status site<svg
          class="text-white align-middle ml-8 relative -top-1 -left-4 transition-all group-hover:left-0"
          style="width: calc(var(--featured-link-icon-size) * 1.25); height: calc(var(--featured-link-icon-size) * 1.25);"
        >
          <use xlink:href="#sprite-icon-gui-link-arrow"></use></svg
      ></a>
    </div>
  </div>
</div>
`;

exports[`Uptime Snapshot Test > react > test case common to both #1`] = `
<div
  class="flex sm:flex-row justify-center items-center p-24 border rounded font-sans font-light text-p2 text-cool-black"
  data-id="uptime-error"
>
  Sorry, we can’t retrieve uptime data right now.
</div>
`;
